### ISSUE TITLE
Memory leak while training

### ISSUE THREAD

**Issue by eaplatanios** [eaplatanios/swift-ale#1]
I tried keeping everything the same while commenting out the `optimizer.update(&network, along: gradient)` line and the leak disappears. This indicates that something may be wrong with the differentiable conformance synthesis (or additive arithmetic, etc.). @rxwei what do you think?

**Comment by eaplatanios** [comment]
To clarify, I actually don't think it goes away. Memory is still leaking but at a much slower pace. This makes me think that something may be off with how we're using the TF eager API.

**Comment by rxwei** [comment]
Hmm, conformance synthesis only produces operations on value types, so a leak is not likely unless something's seriously wrong with Sema and SILGen. 

It's been difficult for me to build swift-ale/retro-swift. Let's merge the protocol split patch first.

**Comment by eaplatanios** [comment]
I have been looking deeper into this and I now know for sure that this is not about auto-diff because even if I remove the gradient computation completely and just do forward passes, I still get a leak. Not sure from where yet. Merging the protocol split patch first sounds good. Currently it seems to blocked by the toolchain used by the CI which has not been updated.

**Comment by rxwei** [comment]
Great. Are you using `class`es anywhere? I may be able to pinpoint potential reference cycles.

**Comment by eaplatanios** [comment]
Actually I noticed something super weird. Consider this loop:
```swift
while true {
  let action: Tensor<Int32> = Raw.multinomial(
    logits: Tensor<Float>(repeating: 1.0, shape: [batchSize, arcadeEnvironment.actionCount]),
    numSamples: Tensor<Int32>(1),
    seed: 0,
    seed2: 0)
  arcadeEnvironment.step(taking: action)
}
```
Changing it to this causes exploding memory usage:
```swift
while true {
  let seed = Context.local.randomSeed
  let action: Tensor<Int32> = Raw.multinomial(
    logits: Tensor<Float>(repeating: 1.0, shape: [batchSize, arcadeEnvironment.actionCount]),
    numSamples: Tensor<Int32>(1),
    seed: Int64(seed.graph),
    seed2: Int64(seed.op))
  arcadeEnvironment.step(taking: action)
}
```
Do you have any idea what could be causing this?

**Comment by eaplatanios** [comment]
@rxwei even simpler example where the memory explodes:
```swift
var i = 0
while true {
  i += 1
  let action: Tensor<Int32> = Raw.multinomial(
    logits: Tensor<Float>(repeating: 1.0, shape: [batchSize, arcadeEnvironment.actionCount]),
    numSamples: Tensor<Int32>(1),
    seed: 0,
    seed2: Int64(i)) // Changing `Int64(i)` to `0` stops the memory from exploding.
  arcadeEnvironment.step(taking: action)
}
```

**Comment by eaplatanios** [comment]
@rxwei reproducer that does not depend on this library or on the protocol split PR:
```swift
var i = 0
var samplesSum = Tensor<Int32>(zeros: [10, 10])
while true {
  i += 1
  samplesSum += Raw.multinomial(
    logits: Tensor<Float>(repeating: 1.0, shape: [10, 10]),
    numSamples: Tensor<Int32>(1),
    seed: 0,
    seed2: Int64(i))
}
```
This will explode pretty fast!

**Comment by eaplatanios** [comment]
Stacktrace:
<img width="1288" alt="Screen Shot 2019-07-26 at 3 53 17 PM" src="https://user-images.githubusercontent.com/1294940/61977829-836ab900-afbd-11e9-8b7b-23b921f62ef1.png">
This may actually be a TF bug.

**Comment by eaplatanios** [comment]
@rxwei I figured it out. I was calling `Raw.multinomial`, but given that we manage our own random seeds in S4TF we should always make sure to use the stateless random ops. Switching to `Raw.statelessMultinomial` fixes the issue. Let's try to keep that  in mind in the future. :P

**Comment by eaplatanios** [comment]
Sorry I'm reopening this because even though the situation is improved, there is still a leak somewhere when training.

**Comment by eaplatanios** [comment]
@rxwei there is definitely something wrong `optimizer.update(&network, along: gradient)`, even when using the `SGD` optimizer. When I uncomment that line, memory will start exploding after a few training steps.

**Comment by eaplatanios** [comment]
Stacktrace:
<img width="1329" alt="Screen Shot 2019-07-26 at 4 44 07 PM" src="https://user-images.githubusercontent.com/1294940/61980387-a3ea4180-afc4-11e9-87a9-8a96f8fc85a3.png">
It does look like something is wrong with `Differentiable.move`.

**Comment by rxwei** [comment]
We'll prioritize it today.

**Comment by rxwei** [comment]
Could you try to compile a derived `Differentiable` struct type and dump the SIL using `-emit-silgen`?

**Comment by eaplatanios** [comment]
Thanks Richard! Is there an easy way to do that for a file that's part of a SwiftPM project and includes dependencies on other projects?

**Comment by rxwei** [comment]
There's no easy way, but I'd suggest creating a standalone differentiable struct that's closest to the one triggering leaks in your library.

**Comment by eaplatanios** [comment]
I tried to create an example, but I haven't tested if the memory leak occurs for this one. In either case `-emit-silgen` outputs 7500 lines...which part are you interested in?

**Comment by eaplatanios** [comment]
The `move` function of that struct looks like this:
```
// ArcadeActorCritic.move(along:)
sil hidden [ossa] @$s5main217ArcadeActorCriticV4move5alongyAC26AllDifferentiableVariablesV_tF : $@convention(method) (@guaranteed ArcadeActorCritic.AllDifferentiableVariables, @inout ArcadeActorCritic) -> () {
// %0                                             // users: %44, %36, %28, %20, %12, %4, %2
// %1                                             // users: %46, %38, %30, %22, %14, %6, %3
bb0(%0 : @guaranteed $ArcadeActorCritic.AllDifferentiableVariables, %1 : $*ArcadeActorCritic):
  debug_value %0 : $ArcadeActorCritic.AllDifferentiableVariables, let, name "direction", argno 1 // id: %2
  debug_value_addr %1 : $*ArcadeActorCritic, var, name "self", argno 2 // id: %3
  %4 = struct_extract %0 : $ArcadeActorCritic.AllDifferentiableVariables, #ArcadeActorCritic.AllDifferentiableVariables.conv1 // user: %5
  %5 = copy_value %4 : $Conv2D<Float>.AllDifferentiableVariables // users: %11, %9
  %6 = begin_access [modify] [unknown] %1 : $*ArcadeActorCritic // users: %10, %7
  %7 = struct_element_addr %6 : $*ArcadeActorCritic, #ArcadeActorCritic.conv1 // user: %9
  // function_ref Conv2D.move(along:)
  %8 = function_ref @$s10TensorFlow6Conv2DV4move5alongyAC26AllDifferentiableVariablesVyx_G_tF : $@convention(method) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Conv2D<τ_0_0>.AllDifferentiableVariables, @inout Conv2D<τ_0_0>) -> () // user: %9
  %9 = apply %8<Float>(%5, %7) : $@convention(method) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Conv2D<τ_0_0>.AllDifferentiableVariables, @inout Conv2D<τ_0_0>) -> ()
  end_access %6 : $*ArcadeActorCritic             // id: %10
  destroy_value %5 : $Conv2D<Float>.AllDifferentiableVariables // id: %11
  %12 = struct_extract %0 : $ArcadeActorCritic.AllDifferentiableVariables, #ArcadeActorCritic.AllDifferentiableVariables.conv2 // user: %13
  %13 = copy_value %12 : $Conv2D<Float>.AllDifferentiableVariables // users: %19, %17
  %14 = begin_access [modify] [unknown] %1 : $*ArcadeActorCritic // users: %18, %15
  %15 = struct_element_addr %14 : $*ArcadeActorCritic, #ArcadeActorCritic.conv2 // user: %17
  // function_ref Conv2D.move(along:)
  %16 = function_ref @$s10TensorFlow6Conv2DV4move5alongyAC26AllDifferentiableVariablesVyx_G_tF : $@convention(method) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Conv2D<τ_0_0>.AllDifferentiableVariables, @inout Conv2D<τ_0_0>) -> () // user: %17
  %17 = apply %16<Float>(%13, %15) : $@convention(method) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Conv2D<τ_0_0>.AllDifferentiableVariables, @inout Conv2D<τ_0_0>) -> ()
  end_access %14 : $*ArcadeActorCritic            // id: %18
  destroy_value %13 : $Conv2D<Float>.AllDifferentiableVariables // id: %19
  %20 = struct_extract %0 : $ArcadeActorCritic.AllDifferentiableVariables, #ArcadeActorCritic.AllDifferentiableVariables.conv3 // user: %21
  %21 = copy_value %20 : $Conv2D<Float>.AllDifferentiableVariables // users: %27, %25
  %22 = begin_access [modify] [unknown] %1 : $*ArcadeActorCritic // users: %26, %23
  %23 = struct_element_addr %22 : $*ArcadeActorCritic, #ArcadeActorCritic.conv3 // user: %25
  // function_ref Conv2D.move(along:)
  %24 = function_ref @$s10TensorFlow6Conv2DV4move5alongyAC26AllDifferentiableVariablesVyx_G_tF : $@convention(method) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Conv2D<τ_0_0>.AllDifferentiableVariables, @inout Conv2D<τ_0_0>) -> () // user: %25
  %25 = apply %24<Float>(%21, %23) : $@convention(method) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Conv2D<τ_0_0>.AllDifferentiableVariables, @inout Conv2D<τ_0_0>) -> ()
  end_access %22 : $*ArcadeActorCritic            // id: %26
  destroy_value %21 : $Conv2D<Float>.AllDifferentiableVariables // id: %27
  %28 = struct_extract %0 : $ArcadeActorCritic.AllDifferentiableVariables, #ArcadeActorCritic.AllDifferentiableVariables.denseHidden // user: %29
  %29 = copy_value %28 : $Dense<Float>.AllDifferentiableVariables // users: %35, %33
  %30 = begin_access [modify] [unknown] %1 : $*ArcadeActorCritic // users: %34, %31
  %31 = struct_element_addr %30 : $*ArcadeActorCritic, #ArcadeActorCritic.denseHidden // user: %33
  // function_ref Dense.move(along:)
  %32 = function_ref @$s10TensorFlow5DenseV4move5alongyAC26AllDifferentiableVariablesVyx_G_tF : $@convention(method) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Dense<τ_0_0>.AllDifferentiableVariables, @inout Dense<τ_0_0>) -> () // user: %33
  %33 = apply %32<Float>(%29, %31) : $@convention(method) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Dense<τ_0_0>.AllDifferentiableVariables, @inout Dense<τ_0_0>) -> ()
  end_access %30 : $*ArcadeActorCritic            // id: %34
  destroy_value %29 : $Dense<Float>.AllDifferentiableVariables // id: %35
  %36 = struct_extract %0 : $ArcadeActorCritic.AllDifferentiableVariables, #ArcadeActorCritic.AllDifferentiableVariables.denseAction // user: %37
  %37 = copy_value %36 : $Dense<Float>.AllDifferentiableVariables // users: %43, %41
  %38 = begin_access [modify] [unknown] %1 : $*ArcadeActorCritic // users: %42, %39
  %39 = struct_element_addr %38 : $*ArcadeActorCritic, #ArcadeActorCritic.denseAction // user: %41
  // function_ref Dense.move(along:)
  %40 = function_ref @$s10TensorFlow5DenseV4move5alongyAC26AllDifferentiableVariablesVyx_G_tF : $@convention(method) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Dense<τ_0_0>.AllDifferentiableVariables, @inout Dense<τ_0_0>) -> () // user: %41
  %41 = apply %40<Float>(%37, %39) : $@convention(method) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Dense<τ_0_0>.AllDifferentiableVariables, @inout Dense<τ_0_0>) -> ()
  end_access %38 : $*ArcadeActorCritic            // id: %42
  destroy_value %37 : $Dense<Float>.AllDifferentiableVariables // id: %43
  %44 = struct_extract %0 : $ArcadeActorCritic.AllDifferentiableVariables, #ArcadeActorCritic.AllDifferentiableVariables.denseValue // user: %45
  %45 = copy_value %44 : $Dense<Float>.AllDifferentiableVariables // users: %51, %49
  %46 = begin_access [modify] [unknown] %1 : $*ArcadeActorCritic // users: %50, %47
  %47 = struct_element_addr %46 : $*ArcadeActorCritic, #ArcadeActorCritic.denseValue // user: %49
  // function_ref Dense.move(along:)
  %48 = function_ref @$s10TensorFlow5DenseV4move5alongyAC26AllDifferentiableVariablesVyx_G_tF : $@convention(method) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Dense<τ_0_0>.AllDifferentiableVariables, @inout Dense<τ_0_0>) -> () // user: %49
  %49 = apply %48<Float>(%45, %47) : $@convention(method) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Dense<τ_0_0>.AllDifferentiableVariables, @inout Dense<τ_0_0>) -> ()
  end_access %46 : $*ArcadeActorCritic            // id: %50
  destroy_value %45 : $Dense<Float>.AllDifferentiableVariables // id: %51
  %52 = tuple ()                                  // user: %53
  return %52 : $()                                // id: %53
} // end sil function '$s5main217ArcadeActorCriticV4move5alongyAC26AllDifferentiableVariablesV_tF'
```

**Comment by rxwei** [comment]
If you have a single-file reproducer that'd be super helpful. You can compile it with `swiftc -g -Xfrontend -gsil` and Instruments will tell you exactly which line of SIL is leaking memory.

**Comment by rxwei** [comment]
> @rxwei even simpler example where the memory explodes:
> 
> ```swift
> var i = 0
> while true {
>   i += 1
>   let action: Tensor<Int32> = Raw.multinomial(
>     logits: Tensor<Float>(repeating: 1.0, shape: [batchSize, arcadeEnvironment.actionCount]),
>     numSamples: Tensor<Int32>(1),
>     seed: 0,
>     seed2: Int64(i)) // Changing `Int64(i)` to `0` stops the memory from exploding.
>   arcadeEnvironment.step(taking: action)
> }
> ```

Could you report this TensorFlow memory leak as an issue on swift-apis? @bgogul and @pschuh are good people to ask.

**Comment by eaplatanios** [comment]
I figured that one out. Ijust had to switch from `Raw.multinomial` to `Raw.statelessMultinomial`. I'm currently looking into whether I can get you a single-file reproducer for the other one, but it looks non-trivial.

**Comment by eaplatanios** [comment]
@rxwei I believe this example is leaking:
```swift
struct ArcadeActorCritic: Module {
  public var dense: Dense<Float> = Dense<Float>(inputSize: 1, outputSize: 1)

  @differentiable(wrt: self)
  public func callAsFunction(_ input: Tensor<UInt8>) -> Tensor<Float> {
    let input = Tensor<Float>(input) / 255.0
    return dense(input)
  }
}

var network = ArcadeActorCritic()
let optimizer = SGD(for: network)
var i = 0
while true {
  let (loss, gradient) = network.valueWithGradient { network  in
    network(Tensor<UInt8>(repeating: 200, shape: [1, 1])).sum()
  }
  optimizer.update(&network, along: gradient)
  if i % 1000 == 0 { print(loss) }
  i += 1
}
```
although at a much smaller scale than the RL example I mentioned.

**Comment by eaplatanios** [comment]
According to instruments, the leaks appear to be happening inside  VJP functions.

**Comment by eaplatanios** [comment]
Actually, both in VJP functions and in `Differentiable.move`.

**Comment by eaplatanios** [comment]
On my machine this example goes from 16MB to about 70MB in ~10 minutes.

**Comment by pschuh** [comment]
@eaplatanios Is there still a leak that involves no AD? or was that just misusing things?

**Comment by eaplatanios** [comment]
I believe there is. Try out this reproducer:
```swift
struct ArcadeActorCritic: Module {
  public var dense: Dense<Float> = Dense<Float>(inputSize: 1, outputSize: 1)

  @differentiable(wrt: self)
  public func callAsFunction(_ input: Tensor<UInt8>) -> Tensor<Float> {
    let input = Tensor<Float>(input) / 255.0
    return dense(input)
  }
}

var network = ArcadeActorCritic()
let optimizer = SGD(for: network)
var i = 0
while true {
  let (loss, gradient) = network.valueWithGradient { network  in
    network(Tensor<UInt8>(repeating: 200, shape: [1, 1])).sum()
  }
  optimizer.update(&network, along: gradient)
  if i % 1000 == 0 { print(loss) }
  i += 1
}
```
It's slowly growing but on my machine it grows from 16MB to about 70MB in ~10 minutes.

**Comment by eaplatanios** [comment]
@pschuh I'm not  sure if this is AD-specific as I also see leaks on `Differentiable.move` as well as  in generated VJPs. This means that all these leaks could possibly be caused by something external to AD.

**Comment by rxwei** [comment]
Looking into it now!

**Comment by rxwei** [comment]
I'm unable to reproduce any memory leak. I made `ArcadeActorCritic` conform to `Layer`. 
<img width="1396" alt="Screen Shot 2019-07-26 at 5 33 11 PM" src="https://user-images.githubusercontent.com/856507/61987439-810f5b80-afcb-11e9-912c-bdcf40d8b0a2.png">

**Comment by rxwei** [comment]
I'll build a toolchain with the newly merged `Module` and try it.

**Comment by eaplatanios** [comment]
Thanks Richard! I am using the new toolchain with `Module` and a `UInt8` input tensor which may play a role I guess.

**Comment by eaplatanios** [comment]
I just tested and with the newest toolchain, even this leaks:
```swift
struct ArcadeActorCritic: Layer {
  public var dense: Dense<Float> = Dense<Float>(inputSize: 1, outputSize: 1)

  @differentiable
  public func callAsFunction(_ input: Tensor<Float>) -> Tensor<Float> {
    return dense(input)
  }
}

var network = ArcadeActorCritic()
let optimizer = SGD(for: network)
var i = 0
while true {
  let (loss, gradient) = network.valueWithGradient { network  in
    network(Tensor<Float>(repeating: 200, shape: [1, 1])).sum()
  }
  optimizer.update(&network, along: gradient)
  if i % 1000 == 0 { print(loss) }
  i += 1
}
```

**Comment by eaplatanios** [comment]
I'm using a toolchain built from the latest commit from apple/swift in the `tensorflow` branch and the version of swift-apis that includes `Module`.

**Comment by eaplatanios** [comment]
@rxwei wow...so if I add this in the same file the leak goes away:
```swift
public protocol Layer: Differentiable, KeyPathIterable
    where TangentVector: VectorProtocol & ElementaryFunctions &
                         PointwiseMultiplicative & KeyPathIterable,
          AllDifferentiableVariables == TangentVector {
    associatedtype Input: Differentiable
    associatedtype Output: Differentiable

    @differentiable
    func callAsFunction(_ input: Input) -> Output
}

@frozen
public struct Dense<Scalar: TensorFlowFloatingPoint>: Layer {
    public var weight: Tensor<Scalar>
    public var bias: Tensor<Scalar>
    @noDerivative public let activation: Activation
    @noDerivative internal let batched: Bool
    public typealias Activation = @differentiable (Tensor<Scalar>) -> Tensor<Scalar>

    init(
        inputSize: Int,
        outputSize: Int,
        activation: @escaping Activation = identity,
        weightInitializer: ParameterInitializer<Scalar> = glorotUniform(),
        biasInitializer: ParameterInitializer<Scalar> = zeros()
    ) {
        self.init(
            weight: weightInitializer([inputSize, outputSize]),
            bias: Tensor(zeros: [outputSize]),
            activation: activation)
    }

    @differentiable
    public func callAsFunction(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
        if batched {
            let hidden = matmul(input.expandingShape(at: 1), weight)
            return activation(hidden.squeezingShape(at: 1) + bias)
        }
        return activation(matmul(input, weight) + bias)
    }
}
```
This indicates that the split to `Module` and `Layer` caused the leak somehow. I'm pretty sure this is a bug in the compiler.

**Comment by eaplatanios** [comment]
@rxwei you don't need to build a new toolchain probably. You may be able to reproduce it by adding the following in the beginning of the file.

```swift
public protocol Module: Differentiable, KeyPathIterable
    where TangentVector: VectorProtocol & ElementaryFunctions &
                         PointwiseMultiplicative & KeyPathIterable,
          AllDifferentiableVariables == TangentVector {
    associatedtype Input
    associatedtype Output: Differentiable

    @differentiable(wrt: self)
    func callAsFunction(_ input: Input) -> Output
}

public protocol Layer: Module where Input: Differentiable {
    @differentiable
    func callAsFunction(_ input: Input) -> Output
}

@frozen
public struct Dense<Scalar: TensorFlowFloatingPoint>: Layer {
    public var weight: Tensor<Scalar>
    public var bias: Tensor<Scalar>
    @noDerivative public let activation: Activation
    @noDerivative internal let batched: Bool
    public typealias Activation = @differentiable (Tensor<Scalar>) -> Tensor<Scalar>

    init(
        inputSize: Int,
        outputSize: Int,
        activation: @escaping Activation = identity,
        weightInitializer: ParameterInitializer<Scalar> = glorotUniform(),
        biasInitializer: ParameterInitializer<Scalar> = zeros()
    ) {
        self.init(
            weight: weightInitializer([inputSize, outputSize]),
            bias: Tensor(zeros: [outputSize]),
            activation: activation)
    }

    @differentiable
    public func callAsFunction(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
        if batched {
            let hidden = matmul(input.expandingShape(at: 1), weight)
            return activation(hidden.squeezingShape(at: 1) + bias)
        }
        return activation(matmul(input, weight) + bias)
    }
}
```

**Comment by rxwei** [comment]
No, I'm unable to compile this.

```swift
import TensorFlow

public protocol Module: Differentiable, KeyPathIterable
    where TangentVector: VectorProtocol & ElementaryFunctions &
                         PointwiseMultiplicative & KeyPathIterable,
          AllDifferentiableVariables == TangentVector {
    associatedtype Input
    associatedtype Output: Differentiable

    @differentiable(wrt: self)
    func callAsFunction(_ input: Input) -> Output
}

public protocol Layer: Module where Input: Differentiable {
    @differentiable
    func callAsFunction(_ input: Input) -> Output
}

@frozen
public struct Dense<Scalar: TensorFlowFloatingPoint>: Layer {
    public var weight: Tensor<Scalar>
    public var bias: Tensor<Scalar>
    @noDerivative public let activation: Activation
    @noDerivative internal let batched: Bool
    public typealias Activation = @differentiable (Tensor<Scalar>) -> Tensor<Scalar>

    init(
        inputSize: Int,
        outputSize: Int,
        activation: @escaping Activation = identity,
        weightInitializer: ParameterInitializer<Scalar> = glorotUniform(),
        biasInitializer: ParameterInitializer<Scalar> = zeros()
    ) {
        self.init(
            weight: weightInitializer([inputSize, outputSize]),
            bias: Tensor(zeros: [outputSize]),
            activation: activation)
    }

    @differentiable
    public func callAsFunction(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
        if batched {
            let hidden = matmul(input.expandingShape(at: 1), weight)
            return activation(hidden.squeezingShape(at: 1) + bias)
        }
        return activation(matmul(input, weight) + bias)
    }
}

struct ArcadeActorCritic: Module {
  public var dense: Dense<Float> = Dense<Float>(inputSize: 1, outputSize: 1)

  @differentiable(wrt: self)
  public func callAsFunction(_ input: Tensor<UInt8>) -> Tensor<Float> {
    let input = Tensor<Float>(input) / 255.0
    return dense(input)
  }
}

var network = ArcadeActorCritic()
let optimizer = SGD(for: network)
var i = 0
while true {
  let (loss, gradient) = network.valueWithGradient { network  in
    network(Tensor<UInt8>(repeating: 200, shape: [1, 1])).sum()
  }
  optimizer.update(&network, along: gradient)
  if i % 1000 == 0 { print(loss) }
  i += 1
}
```

What toolchain are you using?

**Comment by eaplatanios** [comment]
You need the PR that allows for conditional differentiability. The toolchain I am using is the latest from the `tensorflow` branch in apple/swift.

**Comment by eaplatanios** [comment]
Sorry that example had a typo. Try this:
```swift
public protocol Module: Differentiable, KeyPathIterable
    where TangentVector: VectorProtocol & ElementaryFunctions &
                         PointwiseMultiplicative & KeyPathIterable,
          AllDifferentiableVariables == TangentVector {
    associatedtype Input
    associatedtype Output: Differentiable

    @differentiable(wrt: self)
    func callAsFunction(_ input: Input) -> Output
}

public protocol Layer: Module where Input: Differentiable {
    @differentiable
    func callAsFunction(_ input: Input) -> Output
}

@frozen
public struct Dense<Scalar: TensorFlowFloatingPoint>: Layer {
    public var weight: Tensor<Scalar>
    public var bias: Tensor<Scalar>
    @noDerivative public let activation: Activation
    @noDerivative internal let batched: Bool
    public typealias Activation = @differentiable (Tensor<Scalar>) -> Tensor<Scalar>

    init(
        inputSize: Int,
        outputSize: Int,
        activation: @escaping Activation = identity,
        weightInitializer: ParameterInitializer<Scalar> = glorotUniform(),
        biasInitializer: ParameterInitializer<Scalar> = zeros()
    ) {
        self.weight = weightInitializer([inputSize, outputSize])
        self.bias = Tensor(zeros: [outputSize])
        self.activation = activation
        self.batched = false
    }

    @differentiable
    public func callAsFunction(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
        if batched {
            let hidden = matmul(input.expandingShape(at: 1), weight)
            return activation(hidden.squeezingShape(at: 1) + bias)
        }
        return activation(matmul(input, weight) + bias)
    }
}

struct ArcadeActorCritic: Module {
  public var dense: Dense<Float> = Dense<Float>(inputSize: 1, outputSize: 1)

  @differentiable(wrt: self)
  public func callAsFunction(_ input: Tensor<UInt8>) -> Tensor<Float> {
    let input = Tensor<Float>(input) / 255.0
    return dense(input)
  }
}

var network = ArcadeActorCritic()
let optimizer = SGD(for: network)
var i = 0
while true {
  let (loss, gradient) = network.valueWithGradient { network  in
    network(Tensor<UInt8>(repeating: 200, shape: [1, 1])).sum()
  }
  optimizer.update(&network, along: gradient)
  if i % 1000 == 0 { print(loss) }
  i += 1
}
```

**Comment by eaplatanios** [comment]
@rxwei were you able to reproduce the memory leak with the last example?

**Comment by rxwei** [comment]
Yes! Working on a fix tonight

**Comment by eaplatanios** [comment]
Great, thanks for confirming! :)

**Comment by eaplatanios** [comment]
@rxwei I believe this may also be related to [these warnings](https://github.com/tensorflow/swift-apis/issues/395) that were introduced when we split `Layer` and indicates that some kind of resolution of `@differentiable` attributes goes wrong.

**Comment by rxwei** [comment]
Debugging update: The leak comes from a `partial_apply` that's generated before applying subset parameters thunks.
```swift
  // function_ref AD__$s10TensorFlow6matmul_10transposed_AcA0A0VyxGAF_SbAFSbtSjRzAA0aB6ScalarRzlF__jvp_src_0_wrt_0_2
  %111 = function_ref @AD__$s10TensorFlow6matmul_10transposed_AcA0A0VyxGAF_SbAFSbtSjRzAA0aB6ScalarRzlF__jvp_src_0_wrt_0_2 : $@convention(thin) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Tensor<τ_0_0>, Bool, @guaranteed Tensor<τ_0_0>, Bool) -> (@owned Tensor<τ_0_0>, @owned @callee_guaranteed (@guaranteed Tensor<τ_0_0>, @guaranteed Tensor<τ_0_0>) -> @owned Tensor<τ_0_0>) // user: %112
***LEAK***  %112 = partial_apply [callee_guaranteed] %111<τ_0_0>() : $@convention(thin) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Tensor<τ_0_0>, Bool, @guaranteed Tensor<τ_0_0>, Bool) -> (@owned Tensor<τ_0_0>, @owned @callee_guaranteed (@guaranteed Tensor<τ_0_0>, @guaranteed Tensor<τ_0_0>) -> @owned Tensor<τ_0_0>)
  // function_ref AD__orig_$s10TensorFlow6matmul_10transposed_AcA0A0VyxGAF_SbAFSbtSjRzAA0aB6ScalarRzlF_$s10TensorFlow0A0VyxGSbADSbA4DIegggo_Ieggygyoo_ADSbADSbA3DIeggo_Ieggygyoo_AA0aB13FloatingPointRzlTR_src_0_wrt_2_jvp_subset_parameters_thunk
  %113 = function_ref @AD__orig_$s10TensorFlow6matmul_10transposed_AcA0A0VyxGAF_SbAFSbtSjRzAA0aB6ScalarRzlF_$s10TensorFlow0A0VyxGSbADSbA4DIegggo_Ieggygyoo_ADSbADSbA3DIeggo_Ieggygyoo_AA0aB13FloatingPointRzlTR_src_0_wrt_2_jvp_subset_parameters_thunk : $@convention(thin) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Tensor<τ_0_0>, Bool, @guaranteed Tensor<τ_0_0>, Bool) -> (@owned Tensor<τ_0_0>, @owned @callee_guaranteed (@guaranteed Tensor<τ_0_0>) -> @owned Tensor<τ_0_0>) // user: %114
  %114 = partial_apply [callee_guaranteed] %113<τ_0_0>() : $@convention(thin) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Tensor<τ_0_0>, Bool, @guaranteed Tensor<τ_0_0>, Bool) -> (@owned Tensor<τ_0_0>, @owned @callee_guaranteed (@guaranteed Tensor<τ_0_0>) -> @owned Tensor<τ_0_0>) // user: %119
  // function_ref _vjpMatmul<A>(_:transposed:_:transposed:)
  %115 = function_ref @$s10TensorFlow10_vjpMatmul_10transposed_AcA0A0VyxG_AF_AFtAFctAF_SbAFSbtAA0aB13FloatingPointRzlF : $@convention(thin) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Tensor<τ_0_0>, Bool, @guaranteed Tensor<τ_0_0>, Bool) -> (@owned Tensor<τ_0_0>, @owned @callee_guaranteed (@guaranteed Tensor<τ_0_0>) -> (@owned Tensor<τ_0_0>, @owned Tensor<τ_0_0>)) // user: %116
***LEAK***  %116 = partial_apply [callee_guaranteed] %115<τ_0_0>() : $@convention(thin) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Tensor<τ_0_0>, Bool, @guaranteed Tensor<τ_0_0>, Bool) -> (@owned Tensor<τ_0_0>, @owned @callee_guaranteed (@guaranteed Tensor<τ_0_0>) -> (@owned Tensor<τ_0_0>, @owned Tensor<τ_0_0>))
  // function_ref AD__orig_$s10TensorFlow6matmul_10transposed_AcA0A0VyxGAF_SbAFSbtSjRzAA0aB6ScalarRzlF_$s10TensorFlow0A0VyxGSbADSbA4DIeggoo_Ieggygyoo_ADSbADSbA3DIeggo_Ieggygyoo_AA0aB13FloatingPointRzlTR_src_0_wrt_2_vjp_subset_parameters_thunk
  %117 = function_ref @AD__orig_$s10TensorFlow6matmul_10transposed_AcA0A0VyxGAF_SbAFSbtSjRzAA0aB6ScalarRzlF_$s10TensorFlow0A0VyxGSbADSbA4DIeggoo_Ieggygyoo_ADSbADSbA3DIeggo_Ieggygyoo_AA0aB13FloatingPointRzlTR_src_0_wrt_2_vjp_subset_parameters_thunk : $@convention(thin) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Tensor<τ_0_0>, Bool, @guaranteed Tensor<τ_0_0>, Bool) -> (@owned Tensor<τ_0_0>, @owned @callee_guaranteed (@guaranteed Tensor<τ_0_0>) -> @owned Tensor<τ_0_0>) // user: %118
  %118 = partial_apply [callee_guaranteed] %117<τ_0_0>() : $@convention(thin) <τ_0_0 where τ_0_0 : TensorFlowFloatingPoint> (@guaranteed Tensor<τ_0_0>, Bool, @guaranteed Tensor<τ_0_0>, Bool) -> (@owned Tensor<τ_0_0>, @owned @callee_guaranteed (@guaranteed Tensor<τ_0_0>) -> @owned Tensor<τ_0_0>) // user: %119
```

**Comment by rxwei** [comment]
Fixed it. Patch coming :)

**Comment by eaplatanios** [comment]
Thank you so much Richard! I am compiling a toolchain to test now.

**Comment by eaplatanios** [comment]
@rxwei I'm sorry to be the bearer of bad news, but even though the leak is fixed for the simplified single-file example I mentioned here, it still exists when trying to run the full Atari experiment. I have posted some more instructions on how to get this setup and working in the readme file. Could you try to see if after installing the prerequisites `swift run` works? This should showcase the still existing leak.

**Comment by eaplatanios** [comment]
@rxwei @pschuh not sure if the remaining leak is AD-related. Check out this log from instruments:

<img width="1025" alt="Screen Shot 2019-07-28 at 1 09 48 PM" src="https://user-images.githubusercontent.com/1294940/62010459-12e8a700-b139-11e9-8e78-078971d05e15.png">

This seems to indicate that it comes from the way we use the TF eager API.

**Comment by eaplatanios** [comment]
So after every single optimization step it looks like exactly 6.12 MBs are leaked, which come from invoking `Dense.add` in the derived `Differentiable.move`. Something similar happens with `Conv2D.add`.

**Comment by eaplatanios** [comment]
Note that this corresponds exactly to the memory footprint of my model parameters which means that a full copy of the model is leaked after every single update.

**Comment by eaplatanios** [comment]
Also, if I replace `optimizer.update(&network, along: gradient)` with `network.move(along: gradient.scaled(by: ...)` in my training loop, the same thing occurs. So I pretty certain that for the experiment I am running `network.move(along:)` leaks a complete copy of `network`.

**Comment by eaplatanios** [comment]
@rxwei the problem has to be with the eager API somehow. I tried providing my own `Differentiable.move` for dense layers, implemented as:
```swift
public mutating func move(along direction: TangentVector) {
  weight = weight + direction.weight
  bias = bias + direction.bias
}
```
And I still get a leak which is the full size of the dense layer from this function. Are we sure there are no bugs with reference counting for tensor handles?

**Comment by eaplatanios** [comment]
Sorry for spamming btw, but I feel a lot of this info may be useful in debugging this.

**Comment by eaplatanios** [comment]
Using Xcode I see that multiple `TFETensorHandle`s are leaking, but so far I have been unable to create a simple single file reproducer.

**Comment by eaplatanios** [comment]
@rxwei why is the following `TFETensorHandle` not being released given that its reference count is -1 (and how come it can be negative)?
<img width="1108" alt="Screen Shot 2019-07-28 at 4 57 12 PM" src="https://user-images.githubusercontent.com/1294940/62012807-ddec4c80-b158-11e9-9b1c-36e27cfed678.png">

**Comment by eaplatanios** [comment]
@rxwei I figured out where it's coming from and it may be AD-related after all. If I avoid capturing `self` in [this closure](https://github.com/eaplatanios/swift-rl/blob/884be1fb8bb6bf7a9b8d5c49c5cc3ee1ae26d51b/Sources/ReinforcementLearning/Agents/PolicyGradientAgents.swift#L483), by only copying the relevant fields to local variables, the leak is avoided. Do you know why the leak exists otherwise and how to fix it? It's weird because `self` is a struct in that case and so capturing it should not be causing issues.

**Comment by rxwei** [comment]
What would help most is if you have a single-file reproducer. The PR I just merged did fix the leak in the earlier single-file reproducer you provided, and the new issue seems orthogonal.

**Comment by eaplatanios** [comment]
Yes, the new issue is indeed orthogonal. I’ll try to get a single file
reproducer ready soon.

**Comment by eaplatanios** [comment]
@rxwei here you go:
```swift
struct ArcadeActorCritic {
  public var dense: Dense<Float>
  public var optimizer: SGD<Dense<Float>>

  internal var step: Int = 0

  public init(dense: Dense<Float>) {
    self.dense = dense
    self.optimizer = SGD(for: dense)
  }

  @inlinable
  public mutating func update(using value: Tensor<Float>) {
    let (loss, gradient) = dense.valueWithGradient { dense -> Tensor<Float> in
      self.step += 1
      return dense(value).sum()
    }
    if step % 1000 == 0 { print(loss) }
    optimizer.update(&dense, along: gradient)
  }
}

let dense = Dense<Float>(inputSize: 1, outputSize: 1)
var network = ArcadeActorCritic(dense: dense)
while true {
  network.update(using: Tensor<Float>(repeating: 0.7, shape: [1, 1]))
}
```
In general, I believe that the example can be described as follows:

- We have a struct `X`.
- It contains a differentiable property `p`.
- It also contains a mutating function `update` which invokes `p.valueWithGradient { p in ... }` inside it and `...` captures `self` (e.g., by accessing other properties of `X`.

In this case, we get a memory leak of `TFETensorHandle`s.

**Comment by rxwei** [comment]
Thanks for the reproducer. Debugging this has been interesting -- Instruments is producing a lot of red herrings. I'm traveling this week but I'll try to take a look when I can. We are prioritizing fixing any memory leak issue.

**Comment by eaplatanios** [comment]
Thanks Richard! Yes I'm sometimes confused by what's reported in instruments. However, I think I've pinned this down to copies of `self` being leaked when `self` is mutated in calls to `valueWithGradient` from within mutating functions of `self`. I hope the way I describe this is not super confusing.

**Comment by rxwei** [comment]
Fixed in https://github.com/apple/swift/pull/26484.

**Comment by eaplatanios** [comment]
Great, thanks a lot Richard! I'll test again once that PR is merged!

**Comment by eaplatanios** [comment]
@rxwei it seems that the leak is still there. Did you try running the simple reproducer I posted earlier? That one is still leaking.

**Comment by rxwei** [comment]
I did, and it was not leaking. Could you please provide a Swift compiler version (`swift --version`)?

**Comment by rxwei** [comment]
This is the test results for the reproducer you provided [here](https://github.com/eaplatanios/swift-ale/issues/1#issuecomment-515803180). It is not leaking. It'd be useful to always confirm the compiler version to make sure you are using the latest compiler revision.
<img width="1313" alt="Screen Shot 2019-08-07 at 8 58 48 AM" src="https://user-images.githubusercontent.com/856507/62638093-9486c480-b8f1-11e9-927d-889dc34dc697.png">

**Comment by eaplatanios** [comment]
This is the version:
```
Swift version 5.1-dev (LLVM 200186e28b, Swift 0bf4460436)
Target: x86_64-apple-darwin18.6.0
```

**Comment by eaplatanios** [comment]
I believe the compiler is outdated you're right. I'll compile a new toolchain now.

**Comment by eaplatanios** [comment]
@rxwei this did indeed fix the memory leak. Thanks so much! :)

However, another bug seems to have been introduced. I created [TF-742](https://bugs.swift.org/browse/TF-742) to track it.

**Comment by eaplatanios** [comment]
Let's track the other bug in [TF-742](https://bugs.swift.org/browse/TF-742). I'll close this given that the memory leaks have now been fixed.
